### DISCUSSION TITLE
Another Call For Quick Testing

### DISCUSSION THREAD

**Discussion by NoMoreFood** [windirstat/windirstat#163] [General]
@AartBluestoke, @BiatuAutMiahn, @Burger23333, @DaveG-SDX, @Max-RM, @Neustradamus, @PC-SPEZIALIST, @RobertK66, @blyamur, @driscoll42, @erikblake, @hueldoeu, @linuxgurugamer, @reukiodo, @rnmerchant, @salvis, @tron2k

Tagging you all since you helped out on the last one, but I'd really appreciate a sanity check on the bugfix / performance update linked below.  Performance changes were mainly related to duplicate detection.  The only main new feature is the 'Largest Files' tab that's been added as well as Korean language support.  Other than that, just looking to verify that it's stable.

https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2025-01-04

Any feedback is appreciated (comments or reactions).  Hope everyone has a great start to their new year.

**UPDATE: Rescanning issue has now been addressed.**

**Comment by BiatuAutMiahn** [comment]
-Started, enabled dup scan, delete some stuff, browse around in normal mode, then rescan whole directory tree causes the main window to loose focus upon clicking while busy.
-Forced back to front with system informer, goes to background and taskbar icon also disappears as well.
-After enabling visibility, the window stays on interaction. Treemap gets greyed out, tree/listview is navigable, but UI is otherwise locked and does not respond to any commands. Also no pacman animation.

**Reply by NoMoreFood** [reply]
@BiatuAutMiahn Thanks for testing.  I will see if I can produce any of that.  What OS are you running?  Are the behaviors reproducible for you?

**Reply by NoMoreFood** [reply]
Update: I was able to reproduce after repeated attempts.  More to come.  Thanks!

**Reply by BiatuAutMiahn** [reply]
> @BiatuAutMiahn Thanks for testing. I will see if I can produce any of that. What OS are you running? Are the behaviors reproducible for you?

Win11 10.0.22631.4602, and your welcome

**Reply by NoMoreFood** [reply]
@BiatuAutMiahn Can you try reproducing on this release?  

https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2025-01-02

**Comment by blyamur** [comment]
Everything is fine, I ran it this way and that way and restarted and started scanning with duplicates again. But then I pressed stop and the update icon and the application closed.

Windows 11 Pro 24H2 26100.2605
WinDirStat.exe (variant zip)

Addition:
it didn't close, but the icon in the taskbar just disappeared. When I continued working, the application reappeared, but without the icon in the taskbar. after a while the icon appeared

**Reply by BiatuAutMiahn** [reply]
I'd check with a tool that can manage windows for an application hidden or not, like https://systeminformer.com/

**Reply by NoMoreFood** [reply]
@blyamur Can you try reproducing on this release?  

https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2025-01-02

**Reply by blyamur** [reply]
> @blyamur Can you try reproducing on this release?
> 
> https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2025-01-02

After driving it around for a bit, I was unable to replicate the problem.

**Comment by Max-RM** [comment]
Just noticed one bug on Windows 10 ARM32.
I can't install it on SD card.
Although if I will install then move exe to SD card it will work fine.

**Reply by NoMoreFood** [reply]
Yeah, this appears to just be a limitation of the WiX toolset and it seems like they barely support ARM the way it is.  Are you also able to just manually type the drive letter and have it work?

**Reply by Max-RM** [reply]
Doesn't work too.

I already have suggested a way how to solve this problem. Just need to archive portable and installer for ARM32/ARM64 in one archive and then post it in releases.

**Reply by NoMoreFood** [reply]
@Max-RM Both WinDirStat.zip and WinDirStat.7z already have the binaries for all four supported architectures that you can download and run without using the MSI.  Were you thinking something different from that?

**Reply by Max-RM** [reply]
Oh ok. Didn't noticed this file in release list.

**Comment by RobertK66** [comment]
Just run some quick scans with  Beta 2.1.2/2025-01-02  on x64 Win 11 24H2

Played around a little with Advanced Options in Admin mode. All worked fine. No issues - other then some minor freezes when massively clicking and cursor playing with entries in the trees/lists or map... (freezes where never longer than some seconds, so for me this looks all ok)

**Comment by reukiodo** [comment]
The crash from 2.1.1 when scanning all local filesystems seems to be fixed in 2.1.2! yay!

**Comment by AartBluestoke** [comment]
good morning, i ran with 2.1.2/2025-01-02 on x64 Win 11 24H2 
- running under normal mode
- it ran at the same speed as before , just over 40 minutes to scan my drive
![2025-01-03 22_17_22-WinDirStat - 18% C__D_](https://github.com/user-attachments/assets/c7188b43-7474-47d8-9da5-c5ac31483395)
![2025-01-03 22_11_03-WinDirStat - 10% C__D_](https://github.com/user-attachments/assets/8d494d4d-2cbf-4df5-84db-01155ad89755)
![2025-01-03 22_50_30-_new 56 - Notepad++](https://github.com/user-attachments/assets/89855632-1cc5-45af-9076-049c0a300595)
BUT
@NoMoreFood - when i pressed the refresh button this morning, it hard crashed - it froze for a few seconds, then the window just blinked out of existence. I didn't even see a white-out "this program is not responding"

**Reply by AartBluestoke** [reply]
Suggestion for deduplication - 
- don't hash any file over 1mb in size unless you have seen the size before, then go back and back-fill missing hashes if necessary.
- set a lower size below which i don't care eg: "only care if a collision could clean up more than 1mb of space" - and assume less than 10 way collisions - that would mean don't hash anything below 100k.

**Reply by NoMoreFood** [reply]
@AartBluestoke This is how the algorithm currently works:

1. Note the size of the item.
2. If more than one item of that size exists, hash the first 128KB; otherwise exit
3. If more than one item of that first 128KB hash exist hash rest of file; otherwise exit
4. If more than one item hash that full file hash, record it as a duplicate in the list

I did not run exhaustive tests, but that first 128KB seemed to differentiate most files that were actually different.

As for the threshold element, you could get that same effect by using the filtering right now.  I will consider doing something separate for duplicate scanning.

**Comment by AartBluestoke** [comment]
I just replicated that crash, in non-dedupe mode - scanned, then just hit 'rescan' - imitate crash.
- i don't currently have google drive open, it's just drive c and d- as shown above

**Reply by NoMoreFood** [reply]
@AartBluestoke Just so we're running the same baseline, can you try this release: 

https://github.com/windirstat/windirstat/releases/tag/beta%2Fv2.1.2%2F2025-01-03

If you're able to reproduce on that, I would really appreciate a dmp file if possible (sent to berns@uwalumni.com).  Usually they get stashed in (C:\Users\<username>\AppData\Local\CrashDumps).

**Reply by NoMoreFood** [reply]
@AartBluestoke Disregard the dump file request. I finally managed to get it to crash one after like 100 tries.  Looks like a threading  / concurrence issue when cleaning up the top files list.  I'll take a look. Thanks!

**Reply by NoMoreFood** [reply]
@AartBluestoke Can you see if you can repro on this version.  Thanks again!

https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2025-01-04

**Comment by rnmerchant** [comment]
I just ran one with my system (5 drives, nothing 
fancy) and it took a few (?10 minutes) to 
complete; then I clicked 'refresh' and an 
immediate crash.  I can send crashdump if you wish.

Richard

At 2025-01-03 05:24 PM, Bryan Berns wrote:

>***@***.*** 
>Disregard the dump file request. I finally 
>managed to get it to crash one after like 100 
>tries. Looks like a threading / concurrence 
>issue when cleaning up the top files list. I'll take a look. Thanks!
>
>­
>Reply to this email directly, 
><https://github.com/windirstat/windirstat/discussions/163#discussioncomment-11730535>view 
>it on GitHub, or 
><https://github.com/notifications/unsubscribe-auth/AYCZSFF6SQ3XMPJBKJQQMKT2I4Z4RAVCNFSM6AAAAABUN3E3LKVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCNZTGA2TGNI>unsubscribe.
>You are receiving this because you were 
>mentioned.Message ID: 
>***@***.***>



______________________________
Richard N. Merchant

tel (QB)  (250) 909-3206 or (mobile) (778) 834-7440

***@***.***


402 Aldous Rd
Qualicum Beach BC V9K 1M1


Goethe: "Treat people as if they were what they 
ought to be and you help them become what they are capable of becoming."
Maharishi Mahesh Yogi:  "Problems or successes, 
they all are the results of our own actions. 
Karma. The philosophy of action is that no one 
else is the giver of peace or happiness. One's 
own karma, one's own actions are responsible to 
come to bring either happiness or success or whatever."

**Reply by NoMoreFood** [reply]
@rnmerchant Can you try the link below.  Thanks for testing / responding:

https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2025-01-04

**Comment by rnmerchant** [comment]
I've tested this using a small amount of fiddling 
(not so detailed as the previous chap).

No crashing at all this time.

Thanks

Richrd

At 2025-01-04 05:10 AM, Bryan Berns wrote:

>***@***.*** Can 
>you try the link below. Thanks for testing / responding:
>
><https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2025-01-04>https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2025-01-04
>
>­
>Reply to this email directly, 
><https://github.com/windirstat/windirstat/discussions/163#discussioncomment-11733241>view 
>it on GitHub, or 
><https://github.com/notifications/unsubscribe-auth/AYCZSFDIAQQB4MGVWDYDMZT2I7MVRAVCNFSM6AAAAABUN3E3LKVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCNZTGMZDIMI>unsubscribe.
>You are receiving this because you were 
>mentioned.Message ID: 
>***@***.***>



______________________________
Richard N. Merchant

tel (QB)  (250) 909-3206 or (mobile) (778) 834-7440

***@***.***


402 Aldous Rd
Qualicum Beach BC V9K 1M1


Goethe: "Treat people as if they were what they 
ought to be and you help them become what they are capable of becoming."
Maharishi Mahesh Yogi:  "Problems or successes, 
they all are the results of our own actions. 
Karma. The philosophy of action is that no one 
else is the giver of peace or happiness. One's 
own karma, one's own actions are responsible to 
come to bring either happiness or success or whatever."

**Comment by NoMoreFood** [comment]
Thanks everyone for the testing effort.  I'm going to go ahead and publish. If anything pops up in the new production version, please post an issue.  Thanks again!

**Comment by AartBluestoke** [comment]
i think there is still substantial performance still available: for substantial parts of the scan i'm capped by either IO or CPU - i've reasonably fast SSD's on this computer - 2x https://www.crucial.com/ssd/p5-plus/ct2000p5pssd8 - sustaining less than  50mb/s means i'm not IO limited, but also i can't see CPU limits ...
![image](https://github.com/user-attachments/assets/fa542a67-6d1f-47ec-b47a-ffbc89631357)
![image](https://github.com/user-attachments/assets/8b7775c5-3c89-48f7-a219-69744d299714)

edit: i've been unable to replicate the crash on refresh. I'll have the completed run in about half an hour to see if there's any performance regression on my ~40 min dedup check.

**Reply by AartBluestoke** [reply]
well that is a pleasant surprise - finished in 9 minutes...
![image](https://github.com/user-attachments/assets/c4f9289c-7d02-4fbf-9399-407ab1d79a57)

edit: perf of drives attached; 
![image](https://github.com/user-attachments/assets/97407bee-3451-422e-8121-e4b085942e2f)
i closed something that was causing memory pressure and sped up by another minute:
![image](https://github.com/user-attachments/assets/10d98e4d-fc01-4bbe-8701-d961f888ec50)
thoughts on slowdown:
1. are you updating the file on-screen stats counter for each file - or only once per frame in a non-blocking way - the spinner is size/file count is counting might quick there :)
2. how are you allocating memory - removing memory pressure shouldn't make that much difference, but if you are repeatedly grabbing a tiny amount of memory for some objects, rather than recycling something while you are running, that could slow things down.

**Reply by NoMoreFood** [reply]
1. The UI updates are passed off via a temporary vector to the UI thread that copies and updates periodically (about every 300ms).
2. There's a persistent 1 MB buffer for thread-local storage that's used for reading the files in chunks.  Only 128KB of this buffer is used for the partial hash.  For duplication purposes, any memory pressure for large chunks beyond that would be internal to the Windows file system and cryptographic libraries.  There are variety of STL containers that hold pointers and size/hash information that will be doing a lot of small allocations, but I don't think I can optimize that too much (and still maintain my sanity within the code).
